electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +264,100 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
+    super().__init__()
+    self.gnd = self.Port(Ground.empty(), [Common])
+    self.pwr_in = self.Port(VoltageSink.empty())  # high-priority higher-voltage source
+    self.pwr_out = self.Port(VoltageSource.empty())
+
+    self.clamp_voltage = self.ArgParameter(clamp_voltage)
+    self.gate_resistor = self.ArgParameter(gate_resistor)
+
+  def contents(self):
+    super().contents()
+
+    output_current_draw = self.pwr_out.link().current_drawn
+    # Pfet
+    self.fet = self.Block(Fet.PFet(
+      drain_voltage=(0, self.pwr_out.link().voltage.upper()),
+      drain_current=output_current_draw,
+      # gate voltage accounts for a possible power on transient
+      gate_voltage=(self.clamp_voltage.upper(), (self.clamp_voltage.upper())),
+    ))
+
+    self.res = self.Block(Resistor(self.gate_resistor))
+    self.diode = self.Block(ZenerDiode(self.clamp_voltage))
+    # PFet gate to res to gnd
+    self.connect(self.fet.gate, self.res.a)
+    self.connect(self.gnd, self.res.b.adapt_to(Ground()))
+    # pwr  going through the PFet
+    self.connect(self.pwr_in, self.fet.source.adapt_to(VoltageSink()))
+    self.connect(self.pwr_out, self.fet.drain.adapt_to(VoltageSource()))
+    # Connectign the diode from the gate to the pwr out side
+    self.connect(self.fet.drain, self.diode.cathode)
+    self.connect(self.fet.gate, self.diode.anode)
+
+
+
+
+class PmosChargerReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+  @init_in_parent
+  def __init__(self, r1_val: RangeLike = 100*kOhm(tol=0.01), r2_val: RangeLike = 100*kOhm(tol=0.01), rds_on: RangeLike =(0, 0.1)*Ohm):
+    super().__init__()
+
+    self.gnd = self.Port(Ground.empty(), [Common])
+    self.pwr_out = self.Port(VoltageSource.empty(),)  # Load
+    self.vbatt = self.Port(VoltageSink.empty(),)  # Battery
+    self.chg = self.Port(VoltageSink.empty(),)  # Charger
+
+    self.r1_val = self.ArgParameter(r1_val)
+    self.r2_val = self.ArgParameter(r2_val)
+    self.rds_on = self.ArgParameter(rds_on)
+
+  def contents(self):
+    super().contents()
+    max_vcharge_voltage = self.chg.link().voltage.upper()
+    max_vcharge_current = self.chg.link().current_drawn.upper()
+    max_vbatt_voltage = self.vbatt.link().voltage.upper()
+    max_vbatt_current = self.vbatt.link().current_limits.upper() # TODO: check if we should use current_limit or current_draw
+    # Create the PMOS transistors and resistors based on the provided schematic
+    self.mp1 = self.Block(Fet.PFet(
+      drain_voltage=(0, max_vcharge_voltage), drain_current=(0, max_vcharge_current),
+      gate_voltage=(- max_vbatt_voltage, max_vbatt_voltage),
+      rds_on=self.rds_on,
+      power=(0, max_vcharge_voltage * max_vcharge_current)
+    ))
+    self.mp2 = self.Block(Fet.PFet(
+      drain_voltage=(0, max_vbatt_voltage), drain_current=(0, max_vbatt_current),
+      gate_voltage=(0, max_vcharge_voltage),
+      rds_on=self.rds_on,
+      power=(0, max_vbatt_voltage * max_vbatt_current)
```

Comment:
The sustained power dissipation should be pretty minimal - the FET should be in its fully on or off state most of the time.
You could bound it by I^2R using the Rds,on spec and battery current.

Reply:
`      power=(0, max_vbatt_current * max_vbatt_current * self.rds_on.upper())`

Reply:
Make sure it accounts for both charging and discharging current

Maybe this:

```python
# need to include 0 for when the switch is off
batt_current = self.vbatt.link().current_drawn.hull(self.chg.link().current_drawn).hull((0, 0))
power = batt_current * batt_current * self.rds_on
```

Reply:
96b6565